Spacer: add custom units for height and width

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
#22956 

## Description
<!-- Please describe what you have changed or added -->
This PR adds custom unit support to the height and width controls of the Spacer block. It pulls in much of the logic for the same from the Cover block's minHeight support.

### Notes
* I added default values for each unit, to make the experience of switching units more smooth. These might need to be tweaked, especially for horizontal spacers.
* There was a min/max specified for px width spacers. These are still respected, but not for other units. This feels reasonable to me in the absence of the RangeControl.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Vertical spacers
_Insert a Spacer block into a post._
- [ ] Try resizing the spacer in the canvas by dragging the handle.
- [ ] Try changing the height in the inspector controls. Click through several different units.
- [ ] Set the unit to something other than px, and then resize the spacer in the canvas. It should automatically reset the unit to px

Horizontal spacers
_To test a horizontal spacer, insert the Navigation block and add a Spacer block to it._
- [ ] Repeat tests with the width in the horizontal spacer

Mobile
Test on mobile simulators with `npm run native start:reset` to start the packager in dev mode, and `npm run native ios`/`npm run native android`. See screenshots.

## Screenshots <!-- if applicable -->
https://user-images.githubusercontent.com/63313398/140412110-f523964d-7872-4b62-b5a3-3bee89de8890.mov

https://user-images.githubusercontent.com/63313398/147010232-63fe35c4-220e-4981-bfee-cb890610f55f.mp4



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stacimc committed:
Add width and height units and use the custom UnitControl

stacimc committed:
Refactor out the ResizableSpacer and fix resizing in the canvas

stacimc committed:
Correctly initialize width and height for horizontal spacer

stacimc committed:
Handle min and max values for px

glendaviesnz commented:
For me the this works in the editor interface as expected, but it isn't saving the selected unit to the attributes:
![spacer](https://user-images.githubusercontent.com/3629020/142792163-0dd50769-79d7-4d91-a69f-1479146a14d1.gif)

glendaviesnz commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -24,6 +31,133 @@ const MAX_SPACER_HEIGHT = 500;
 const MIN_SPACER_WIDTH = 1;
 const MAX_SPACER_WIDTH = 500;
 
+function DimensionInput( {
```
Would it be worth exporting these smaller components and adding some unit tests? Also, I know it is common to have multiple functional components in a single file, but I wonder it is worth trying to move more to a single file -> single component approach? I don't have a strong opinion on it, but it does seem that some of the core block files can get quite long and complicated.

stacimc committed:
Apply unit to the dimensions on the frontend

stacimc committed:
Refactor out the controls

stacimc commented:
Thanks for testing, @glendaviesnz! This has been rebased and updated to respect units on the frontend :)

> I wonder it is worth trying to move more to a single file -> single component approach? I don't have a strong opinion on it, but it does seem that some of the core block files can get quite long and complicated.

There's a lot that could be done here. I think it might be a good idea to make a `ResizableBoxWithOrientation` component and use it both here and in the `Cover` block 🤔 For now for a start, I've broken out the controls into a separate file. I'll look into refactoring further.

jasmussen commented:
Thanks for this PR. Additional unit support has been requested quite a few times, and I think you'll make a lot of people happy with landing this. I took it for a spin, and it worked rather well for everything but percentage units:
![vertical](https://user-images.githubusercontent.com/1204802/144801978-1ce6e2d3-f2be-4aa7-8389-8cb40dc25c38.gif)

Do you know what's up with percentage units that's making that one tricky? My best guess: the problem is that percentages are relative to the parent, making it hit or miss whether percentages work at all. I.e. if it's inside a parent with zero height, the spacer will be zero height. In that light, can we disable just the percentage unit from spacers? Even doing so in a UI-only way (not preventing themers from still defining them in the code if they want to) would be fine.

kjellr commented:
> In that light, can we disable just the percentage unit from spacers? Even doing so in a UI-only way (not preventing themers from still defining them in the code if they want to) would be fine.

This seems fine to me. 👍

stacimc committed:
Disable percentage units for Spacer

stacimc committed:
Only output width or height styles as needed for orientation

stacimc committed:
Correctly generate styles when width or height is set to 0

stacimc commented:
> In that light, can we disable just the percentage unit from spacers? 

Thanks for catching this! I think disabling the percentage unit is a good idea, as the majority of the time they will not be in a parent container where this makes sense/is meaningful. Updated.

jasmussen reviewed (approved):
Took this for a good spin, and in my testing it works very well, with no perceptible downsides. It works vertically and horizontally and the units appear to behave as they should:
![spacer](https://user-images.githubusercontent.com/1204802/144991140-a3a17a7e-02c0-4614-85a4-94dd7c5a6b94.gif)

The omission of percentage makes complete sense to me, and feels like a good solution.

I'd encourage a sanity check on the code, but overall, this feels solid. Thank you for your work!

jasmussen commented:
Looks like the snapshots need to be updated.

kjellr commented:
Question: Is it possible to use dynamic CSS values like variables or `min()`/`max()` if we adopt this approach? That's possible for many other areas that accept custom units (font size, margin, padding, etc.). But because the markup here requires a `heightUnit` attribute, I'm not sure how that would work.

The main use case I have in mind is [this space beneath the header in Twenty Twenty-Two](https://github.com/WordPress/gutenberg/pull/37105). I'd love to use a spacer to replace that, but to keep it in sync with other areas of the theme we currently use a `max(1.25rem, 5vw)` value so that the space isn't too large on small screens. 

This wouldn't need to be visible in the UI, but if there were a "custom" unit size or something that would allow other values that would be great.

Mamaduka commented on packages/block-library/src/spacer/controls.js:
```diff
@@ -0,0 +1,156 @@
+} ) {
+	const [ temporaryInput, setTemporaryInput ] = useState( null );
+
+	const instanceId = useInstanceId( UnitControl );
+	const inputId = `block-spacer-height-input-${ instanceId }`;
```
A prefix can be passed as a second argument for `useInstanceId`.

stacimc replied:
Good to know! Updated.

Mamaduka commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -97,88 +143,64 @@ const SpacerEdit = ( {
-			updateWidth( 72 );
-			updateHeight( 0 );
+			setAttributes( {
+				height: parseFloat( 0 ),
```
Is there a reason for parsing hardcoded primitive values?

stacimc replied:
Definitely not -- fixed.

Mamaduka commented:
Hi, @stacimc

Thanks for working on this 🌟 

What's the reason for storing unit values as separate attributes?

stacimc committed:
Remove unnecessary parsing, interpolation

stacimc commented:
> What's the reason for storing unit values as separate attributes?

@Mamaduka ~~The units were added as separate attributes to prevent needing to change the attribute type for the existing `height` and `width` attributes, which would invalidate existing blocks.~~  It also makes it easier to do some of the manipulation when dragging to resize the spacer without needing to parse out the unit -- for example, updating to `px` units when the handle is dragged. It's the same approach that was taken when adding custom units to Cover's minHeight (#20888).

@kjellr brings up an interesting limitation of using separate attributes. 

Edit: I may have been incorrect in assuming that changing the attribute type would require a deprecation. I'll play around with this a bit more.

stacimc committed:
Remove separate unit attributes and update type of height and width attrs

stacimc committed:
Add block deprecation

stacimc committed:
Update and add fixtures for deprecation

stacimc committed:
Update snapshots

stacimc commented:
This has been updated to remove the `heightUnit` and `widthUnit` attributes, instead changing `height` and `width` to strings in order to support complex values as well. I've also added a deprecation to normalize the values, and added/updated fixtures and snapshots.

@kjellr I think this should work for the complex value case you brought up 🤞 I've tested pasting in the following:

```
<!-- wp:spacer {"height":"max(1.25rem, 5vw)"} -->
<div style="height:max(1.25rem, 5vw)" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->
```

This works as you would expect in different screen sizes 😄 

Note: It is not possible to add or edit complex vals through the panel UI, so if you select this Spacer in the editor, the panel displays an empty input. I've opened #37303 to discuss updating the UI to make it possible to edit custom values in the panel as well.

What do you think?

stacimc committed:
Ensure that the unit is updated to px when resizing by dragging the handle

kjellr commented:
Cool, thanks! I'll give this a test later today. 

> Note: It is not possible to add or edit complex vals through the panel UI, so if you select this Spacer in the editor, the panel displays an empty input. I've opened #37303 to discuss updating the UI to make it possible to edit custom values in the panel as well.

That's no problem at this point. Most users don't need to be able to edit those values, just replace them. There's an issue here to discuss how to better display these values to users: https://github.com/WordPress/gutenberg/issues/35558

kjellr reviewed (approved):
This could use a final code review, but from a design + function perspective, this works well for me!

ramonjd commented:
Just swinging by to give a test. This works well and I think it's good to go as others have noted.

I'm not sure this is feasible, but one nice to have in my opinion might be to add a width of 100% to the vertical spacer container so that it's accessible in vertical navigation menus.

For me it appears with zero width, which makes manipulating it via the handles tough:

![Kapture 2021-12-16 at 09 42 01](https://user-images.githubusercontent.com/6458278/146277250-7408d4c2-2da4-43d5-be1f-545f720798df.gif)

Not a blocker, as the spacer can be still resized using the input control.

aristath commented:
Thank you @stacimc for this PR, looks good and works as expected. 👍 
I see there are quite a few failing tests right now, so I rebased the PR since it was behind trunk by a lot. Let's see if tests pass now, and if they do this should be good to merge 👍

stacimc commented:
Investigating the breaking tests on mobile.  👀  I'll look into updating mobile support.

stacimc committed:
Add mobile support for custom units

stacimc commented:
Updated with mobile support for the custom units.

aaronrobertshaw reviewed (approved):
Nice work @stacimc 👍 

As others have mentioned, this is testing well. I've just given it another run after the latest updates for Native.

✅ Fixture tests pass locally
✅ Vertical spacer:
    - Can resize spacer via dragging the handle
    - Changing the height via inspector controls works across units
    - Inspector control reflects unit change to `px` when dragging handles with prior non-px unit
✅ Horizontal spacer:
    - Can resize spacer via dragging the handle
    - Changing the width via inspector controls works across units
    - Inspector control reflects unit change to `px` when dragging handle with prior non-px unit
✅ Native:
    - Can adjust the spacer height
    - Switching units appears to work

❓ On Native, the step value for `em` and `rem` seems overly small at `0.01`. Can this be improved? Obviously not a blocker and can be addressed separately if the tweak is desired.

![Spacer-Native](https://user-images.githubusercontent.com/60436221/147051105-e6787201-5d79-4d20-beed-422ea5ece305.gif)

aaronrobertshaw commented:
Also, I hope you don't mind but I rebased the PR branch again in the hope that picking up recent commits such as https://github.com/WordPress/gutenberg/pull/37501 will get the failing e2e test passing.

stacimc commented:
> Also, I hope you don't mind but I rebased the PR branch again in the hope that picking up recent commits such as #37501 will get the failing e2e test passing.

Thank you!

>  On Native, the step value for em and rem seems overly small at 0.01. Can this be improved? Obviously not a blocker and can be addressed separately if the tweak is desired.

The step value comes from the configuration [here](https://github.com/WordPress/gutenberg/blob/2e5af3fc1f2082eabe0f113e728b18d7d3b4703f/packages/components/src/unit-control/utils.ts#L34). It's the same for the Cover block `minHeight` for example, and I think it makes the increment look slow/too small on desktop as well:

https://user-images.githubusercontent.com/63313398/147142811-1863b868-0aea-4ea4-adcc-3830db7f799a.mov


Investigating, it seems like different steps were chosen for each category of units, with `em`/`rem` as "character-size based units" (#32692) If we change that step in the configuration, we could update it everywhere at once. From a UI perspective I would be on board with that change, although maybe in a separate PR since it touches more than just Spacer, and to allow others to weigh in. What do you think?

aaronrobertshaw commented:
> From a UI perspective I would be on board with that change, although maybe in a separate PR since it touches more than just Spacer, and to allow others to weigh in. What do you think?

That was my thinking, to address it in a separate PR. It could be a little subjective as well. 

Another perspective to consider is that most using the step functionality will be doing so to fine-tune a value rather than set it for these units. The small step value makes more sense in this case.

stacimc commented:
> That was my thinking, to address it in a separate PR. It could be a little subjective as well.

Sounds good. I opened #37620 to allow for discussion.

Merging, but I'm not clear if this should receive a `Backport` label -- @kjellr or @jasmussen can you weigh in here on whether the use case warrants backporting?

kjellr commented:
👋 I'm unsure if it requires back porting. We haven't used this in Twenty Twenty-Two at this point, so it's not _technically_ required. Though we do have https://github.com/WordPress/twentytwentytwo/issues/272, which could technically be solved by this. I'll take a closer look in the next day or so.

kjellr commented:
I opened https://github.com/WordPress/twentytwentytwo/pull/319 to use these custom units in some of Twenty Twenty-Two's patterns and templates. This fixes a bug we had with the header template margin not being editable, and I think it's a very noticeable improvement. So let's try and backport this for release. 

cc @noisysocks.

noisysocks commented:
We're past RC 1 unfortunately so not able to land new functionality. If this isn't _technically_ required, like you say, then It think let's leave it for WP 5.9.1.

kjellr commented:
@noisysocks I think this one is a bit of a special case — it's technically a new feature here in Gutenberg, but it fixes a somewhat obvious bug in Twenty Twenty-Two: the fact that you can't edit the space here via the UI: 

Site Editor|Template Editor
---|---
<img width="1408" alt="Screen Shot 2022-01-05 at 7 45 51 AM" src="https://user-images.githubusercontent.com/1202812/148219909-568fbe99-6256-4a7f-b68e-a08557d6bb3a.png">|<img width="1408" alt="Screen Shot 2022-01-05 at 7 49 50 AM" src="https://user-images.githubusercontent.com/1202812/148220417-d7005ec4-e88a-4b8f-87b8-f8377fc2c83b.png">


If this were something we could fix in the theme, I'd do it, but unfortunately it's not (at least, not in a way that works well for mobile + desktop). We tried to fix it before in https://github.com/WordPress/gutenberg/pull/37105, but decided to continue pursue the approach in this PR instead. 

I understand why we'd avoid pushing it, but I think it's worth a thorough consideration at least. That space is one of the first things a user is going to see when they activate Twenty Twenty-Two, and it's rough that there's no clear way to edit it.

Mamaduka commented:
If we decide to backport this, we should also include #37774.

noisysocks commented:
Sorry, it's too late. There's a bit of new functionality here which puts it more towards the enhancement end of the spectrum. Also there are new strings which would not be localised in time as we're now past the RC 1 hard string freeze.

We can include it in 5.9.1. We usually do a quick minor release after releasing a version with so many substantial changes (e.g. 5.0.1 and 5.0.2 came pretty quick after 5.0).

kjellr commented:
👍 No worries, thanks.

Mamaduka commented:
Cherry picked for 5.9.1.
